### PR TITLE
[serverless][SLS-2732] Fix runtime duration and post runtime duration values

### PR DESCRIPTION
**Pull request by DylanLovesCoffee**
### What does this PR do?

TLDR: Updates the Extension to avoid calculating runtime duration (for the `REPORT` log) and post runtime duration (for the `REPORT` log and the enhanced metric) when the invocation start or end time is a zero value.

Previously, we were assuming that when we process the `platform.report` log, we can calculate the runtime duration and post runtime duration, since we need to process the `platform.runtimeDone` and `platform.start` log messages first to set an `invocationStart|EndTime`. However, some of the times (especially during error invocations like OOM) we get these logs in different orders.

### Motivation

Without these checks in place, we report durations that are incredibly large.

Report logs:
Screenshot 2023-08-21 at 11 27 15 AM

Post Runtime Metrics:
Screenshot 2023-08-21 at 2 05 52 PM

### Describe how to test/QA your changes

Updated tests + manual testing

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/serverless/metrics/enhanced_metrics.go [hunk 1]
```diff
@@ -188,14 +188,21 @@ func GenerateEnhancedMetricsFromReportLog(args GenerateEnhancedMetricsFromReport
...
 	})
-	args.Demux.AggregateSample(metrics.MetricSample{
-		Name:       postRuntimeDurationMetric,
-		Value:      postRuntimeDuration,
-		Mtype:      metrics.DistributionType,
-		Tags:       args.Tags,
-		SampleRate: 1,
-		Timestamp:  timestamp,
-	})
+	if args.RuntimeStart.IsZero() || args.RuntimeEnd.IsZero() {
```
Comment (IvanTopolcic) [inline pkg/serverless/metrics/enhanced_metrics.go:191]: I wonder if we could keep a running tail of all of the requests instead of defaulting to not calculating it. For instance, whenever we get the REPORT/END log lines, we could add them to a global map, and then calculate them at that point instead.

**Review by IvanTopolcic — APPROVED** [review]
LGTM, just a little nit + some thoughts for the future

File: pkg/serverless/metrics/enhanced_metrics.go (hunk 1 above)
Comment (DylanLovesCoffee) [reply pkg/serverless/metrics/enhanced_metrics.go:191]: I think I get what you mean. So we map `reqId` -> `start|runtimeDone` logs, and only when we have both items then we calculate the runtime duration metrics

File: pkg/serverless/metrics/enhanced_metrics.go [hunk 2]
```diff
@@ -188,14 +188,21 @@ func GenerateEnhancedMetricsFromReportLog(args GenerateEnhancedMetricsFromReport
...
+		fmt.Printf("args.RuntimeStart: %v\n", args.RuntimeStart)
+		fmt.Printf("args.RuntimeEnd: %v\n", args.RuntimeEnd)
```
[the remark this answers is not in the document]
Comment (DylanLovesCoffee) [reply pkg/serverless/metrics/enhanced_metrics.go:193]: 🙈

File: pkg/serverless/metrics/enhanced_metrics.go (hunk 1 above)
Comment (IvanTopolcic) [reply pkg/serverless/metrics/enhanced_metrics.go:190]: Yes, exactly. Does that sound like a good idea to you? There's probably some considerations (i.e, we may want to expire requests after a while if we don't receive a runtimeDone log) but besides that I think it's fairly reasonable.

Comment (DylanLovesCoffee) [reply pkg/serverless/metrics/enhanced_metrics.go:190]: 👍 yup that makes perfect sense. I'll log a card for us
